File: typescript/bin/migrate-create.mjs
#!/usr/bin/env node
// Scaffolds a new <NNNN>_<name>_{up,down}.sql pair in the migrations directory. Numbering is max(1-4-digit prefix in dir)+1; legacy timestamp prefixes are ignored.

import {  mkdir, readdir, writeFile  } from "node:fs/promises";
import { normalizeDialect } from "../dist/index.js";

const NAME_RE = /^[a-z][a-z0-9_]*$/;
const SEQ_RE = /^(\d{1,4})_.*_up\.sql$/;

const HELP_TEXT = `Usage: migrate-create --provider <sqlite|postgres|mysql|sqlserver|oracle> --name <snake_case_slug> [--migrations-path <dir>]

Creates a new <NNNN>_<name>_up.sql + <NNNN>_<name>_down.sql pair scaffolded
with TODO placeholders. NNNN is one greater than the highest 1-4-digit
sequence prefix already present in the migrations directory (starts at 0001).
Legacy timestamp-prefixed filenames are ignored for numbering.

  --provider          Database dialect. Required.
  --name              snake_case slug for the new migration. Required. Must
                      match /^[a-z][a-z0-9_]*$/.
  --migrations-path   Migrations directory (default: ./sql/<dialect>/migrations).
                      Created with mkdir -p if it does not exist.

Examples:
  migrate-create --provider sqlite --name add_users
  migrate-create --provider postgres --name backfill_email_lower --migrations-path ./db/changes/postgres
`;

function parseArgs(argv) {
  const args = { provider: null, name: null, migrationsPath: null };
  const rest = argv.slice(2);
  for (let i = 0; i < rest.length; i++) {
    const a = rest[i];
    const take = (flag) => {
      const eq = `${flag}=`;
      if (a === flag) return rest[++i];
      if (a.startsWith(eq)) return a.slice(eq.length);
      return undefined;
    };
    if (a === "-h" || a === "--help") {
      process.stderr.write(HELP_TEXT);
      process.exit(0);
    }
    let v;
    if ((v = take("--provider")) !== undefined) args.provider = v;
    else if ((v = take("--name")) !== undefined) args.name = v;
    // --migrate-path is a silent alias retained for backward compat with pre-rename callers.
    else if ((v = take("--migrations-path")) !== undefined)
      args.migrationsPath = v;
    else if ((v = take("--migrate-path")) !== undefined)
      args.migrationsPath = v;
    else {
      process.stderr.write(`unknown arg: ${a}\n`);
      process.exit(2);
    }
  }
  return args;
}

async function nextSequence(dir) {
  let entries;
  try {
    entries = await readdir(dir);
  } catch (err) {
    if (err.code === "ENOENT") return 1;
    throw err;
  }
  let max = 0;
  for (const name of entries) {
    const m = SEQ_RE.exec(name);
    if (!m) continue;
    const n = Number.parseInt(m[1], 10);
    if (n > max) max = n;
  }
  return max + 1;
}

function pad4(n) {
  return String(n).padStart(4, "0");
}

function main() {
  const args = parseArgs(process.argv);
  if (!args.provider) {
    process.stderr.write("missing --provider\n");
    process.exit(2);
  }
  const dialect = normalizeDialect(args.provider);
  if (!dialect) {
    process.stderr.write(`unknown provider: ${args.provider}\n`);
    process.exit(2);
  }
  if (!args.name) {
    process.stderr.write("missing --name\n");
    process.exit(2);
  }
  if (!NAME_RE.test(args.name)) {
    process.stderr.write(
      `--name must be snake_case: lowercase letter then [a-z0-9_] (got "${args.name}")\n`,
    );
    process.exit(2);
  }

  const migrationsPath = args.migrationsPath ?? `./sql/${dialect}/migrations`;
  await mkdir(migrationsPath, { recursive: true });

  const seq = pad4(await nextSequence(migrationsPath));
  const upName = `${seq}_${args.name}_up.sql`;
  const downName = `${seq}_${args.name}_down.sql`;
  await writeFile(
    `${migrationsPath}/${upName}`,
    `-- TODO: write the up migration for "${args.name}"\n`,
  );
  await writeFile(
    `${migrationsPath}/${downName}`,
    `-- TODO: write the down migration for "${args.name}"\n`,
  );
  process.stdout.write(`Created: ${upName}\n`);
  process.stdout.write(`Created: ${downName}\n`);
}

try {
  main();
} catch (e) {
  process.stderr.write(`${e?.stack ?? e?.message ?? String(e)}\n`);
  process.exit(1);
}

File: rust/src/bin/migrate_create.rs
// Scaffolds a new <NNNN>_<name>_{up,down}.sql pair in the migrations directory. Numbering is max(1-4-digit prefix in dir)+1; legacy timestamp prefixes are ignored.

use std::env;
use std::fs;
use std::path::PathBuf;
use std::process;

use deterministic_migraters::{fill_help_template, message, scaffold, try_get, SUPPORTED_PROVIDERS};

const HELP_TEMPLATE: &str = include_str!("../../../shared/templates/help/create.txt");

fn help_text() -> String {
    fill_help_template(HELP_TEMPLATE, "migrate-create")
}

struct Args {
    name: String,
    migrations_path: String,
}

fn parse_args() -> Result<Args, String> {
    let mut provider: Option<String> = None;
    let mut name: Option<String> = None;
    let mut migrations_path: Option<String> = None;
    let mut iter = env::args().skip(1);
    while let Some(a) = iter.next() {
        match a.as_str() {
            "--provider" => provider = iter.next(),
            "--name" => name = iter.next(),
            // --migrate-path is a silent alias retained for backward compat.
            "--migrations-path" => migrations_path = iter.next(),
            "--migrate-path" => migrations_path = iter.next(),
            "-h" | "--help" => {
                eprint!("{}", help_text());
                process::exit(0);
            }
            other => return Err(message("errors/unknown-arg", &[("arg", other)])),
        }
    }
    let Some(provider) = provider else {
        return Err(message(
            "errors/missing-provider-hint",
            &[("providers", SUPPORTED_PROVIDERS)],
        ));
    };
    let Some(dialect) = try_get(&provider) else {
        return Err(message(
            "errors/unsupported-provider",
            &[("provider", &provider)],
        ));
    };
    let Some(name) = name else {
        return Err(message("errors/missing-name-hint", &[]));
    };
    if !is_snake_case(&name) {
        return Err(message("errors/invalid-name", &[("name", &name)]));
    }
    let migrations_path =
        migrations_path.unwrap_or_else(|| format!("./sql/{}/migrations", dialect.name()));
    Ok(Args {
        name,
        migrations_path,
    })
}

fn is_snake_case(s: &str) -> bool {
    let mut chars = s.chars();
    match chars.next() {
        Some(c) if c.is_ascii_lowercase() => {}
        _ => return false,
    }
    chars.all(|c| c.is_ascii_lowercase() || c.is_ascii_digit() || c == '_')
}

// Bounded to 1-4 digits so legacy timestamp prefixes (10+ digits, e.g. 1767261600_initial_up.sql) do not poison numbering.
fn parse_sequence(filename: &str) -> Option<u32> {
    if !filename.ends_with("_up.sql") {
        return None;
    }
    let bytes = filename.as_bytes();
    let mut end = 0usize;
    while end < bytes.len() && end < 4 && bytes[end].is_ascii_digit() {
        end += 1;
    }
    if end == 0 {
        return None;
    }
    if end >= bytes.len() || bytes[end] != b'_' {
        return None;
    }
    // Reject 5+ digit prefixes (the 5th char must NOT be a digit).
    if end == 4 && bytes.len() > 4 && bytes[4].is_ascii_digit() {
        return None;
    }
    filename[..end].parse::<u32>().ok()
}

fn next_sequence(dir: &str) -> Result<u32, Box<dyn std::error::Error>> {
    let path = PathBuf::from(dir);
    if !path.is_dir() {
        return Ok(1);
    }
    let mut max = 0u32;
    for entry in fs::read_dir(&path)? {
        let entry = entry?;
        let name = match entry.file_name().into_string() {
            Ok(s) => s,
            Err(_) => continue,
        };
        if let Some(n) = parse_sequence(&name) {
            if n > max {
                max = n;
            }
        }
    }
    Ok(max + 1)
}

fn main() {
    let args = match parse_args() {
        Ok(a) => a,
        Err(msg) => {
            eprintln!("{}", msg);
            process::exit(2);
        }
    };
    if let Err(e) = run(&args) {
        eprintln!("{}", e);
        process::exit(1);
    }
}

fn run(args: &Args) -> Result<(), Box<dyn std::error::Error>> {
    let dir = PathBuf::from(&args.migrations_path);
    fs::create_dir_all(&dir)?;
    let seq = next_sequence(&args.migrations_path)?;
    let prefix = format!("{:04}", seq);
    let up_name = format!("{}_{}_up.sql", prefix, args.name);
    let down_name = format!("{}_{}_down.sql", prefix, args.name);
    let name = args.name.as_str();
    fs::write(dir.join(&up_name), scaffold("up", &[("name", name)]))?;
    fs::write(dir.join(&down_name), scaffold("down", &[("name", name)]))?;
    print!("{}", message("status/created", &[("file", &up_name)]));
    print!("{}", message("status/created", &[("file", &down_name)]));
    Ok(())
}
